onnxruntime/core/graph/graph.cc

```diff
@@ -1047,6 +1050,14 @@ Graph::Graph(const Model& owning_model,
 #endif
   }
 
+  // Ensure initializers are also graph inputs.
+  if (ir_version_ < 4) {
+    for (const auto* initializer : created_initializers) {
+      ValueInfoProto* input = graph_proto_->add_input();
+      utils::TensorProtoToValueInfoProto(*initializer, *input);
+    }
+  }
+
```

Comment:
Isn't this 'fixing' an invalid graph by adding the initializers as graph inputs? 

Can you elaborate on how/when this issue is seen? Is it due to some issue with serialization to ONNX from onnxruntime::Graph or Node when IR version < 4? If that's the case I would have expected the fix to be in Graph::ToGraphProtoInternal or Node::ToProto (if the issue is with a subgraph).

Reply:
Thanks for taking a look @skottmckay 

* There's an ONNX model with a subgraph.
* The subgraph contains constant nodes.
* The model has `ir_version: 3`
* ONNX checker says the model is valid.
* ORT internally changes the constant nodes into Graph initializers.
* But it does not add them as graph inputs, which violates the restriction in `ir_version < 4`.

The test case I added illustrates this. It fails w/o this fix and passes with it.

I'm happy to make the change wherever you think is best, but it seemed quite logical to do it at the point where the initializers are being added.

Reply:
I would have expected this code to put the initializers on the list of graph inputs:
https://github.com/microsoft/onnxruntime/blob/c579ebfbc3f07aa5636093ef033ff4187074a7d1/onnxruntime/core/graph/graph.cc#L3659-L3661

We use that list when we write out the model using Graph::ToGraphProto, so converting a Constant node to an initializer should be handled. 

https://github.com/microsoft/onnxruntime/blob/c579ebfbc3f07aa5636093ef033ff4187074a7d1/onnxruntime/core/graph/graph.cc#L3390-L3392

When writing out a subgraph we either preserve the original, or use Graph::ToGraphProto so that should take the same path. 

https://github.com/microsoft/onnxruntime/blob/c579ebfbc3f07aa5636093ef033ff4187074a7d1/onnxruntime/core/graph/graph.cc#L572-L575

Based on that I am missing what path your model loading/saving is taking that results in graph_inputs_including_initializers_ being invalid. Is Graph::SetInputs being called?

Reply:
Actually when we load from a GraphProto we don't do the input fixup logic, so line 3659 isn't hit. 

Is it enough to just add the initializer to graph_proto_.input when creating the initializer from a Constant node if IR version is < 4? May be able to do that by calling GetOrCreateNodeArg followed by NodeArg::ToProto to get the ValueInfoProto using existing code.

Reply:
I tried to implement what I think you suggested, but it wasn't entirely obvious to me. If this isn't acceptable, can we schedule a quick meeting to discuss?

Reply:
Changes look fine. There's a helper that can convert TensorProto to TypeProto to avoid duplicating that code. 

Any reason to not add the graph input immediately after utils::ConstantNodeProtoToTensorProto returns successfully so we don't need the created_initializers vector?

e.g. 

```cpp
    const gsl::not_null<TensorProto*> tensor{graph_proto_->add_initializer()};
    auto status = utils::ConstantNodeProtoToTensorProto(node, model_path, *tensor);
    ORT_ENFORCE(status.IsOK(), status.ToString());

    // Ensure initializers are also graph inputs.
    if (ir_version_ < 4) {
      TypeProto t{TypeProtoFromTensorProto(*tensor)};
      const NodeArg& node_arg = GetOrCreateNodeArg(tensor->name(), &t);
      *(graph_proto_->add_input()) = node_arg.ToProto();
    }
```

Reply:
Done.